WordPressAuthenticator/GoogleSignIn/NewGoogleAuthenticator.swift

```diff
@@ -0,0 +1,97 @@
+import AuthenticationServices
+
+public class NewGoogleAuthenticator: NSObject {
+
+    let clientId: GoogleClientId
+    let scheme: String
+    let audience: String
+    let contextProvider: ASWebAuthenticationPresentationContextProviding
+
+    let oauthTokenGetter: GoogleOAuthTokenGetting
+
+    public convenience init(
+        clientId: GoogleClientId,
+        scheme: String,
+        audience: String,
+        contextProvider: ASWebAuthenticationPresentationContextProviding,
+        urlSession: URLSession
+    ) {
+        self.init(
+            clientId: clientId,
+            scheme: scheme,
+            audience: audience,
+            contextProvider: contextProvider,
+            oautTokenGetter: GoogleOAuthTokenGetter(dataGetter: urlSession)
+        )
+    }
+
+    init(
+        clientId: GoogleClientId,
+        scheme: String,
+        audience: String,
+        contextProvider: ASWebAuthenticationPresentationContextProviding,
+        oautTokenGetter: GoogleOAuthTokenGetting
+    ) {
+        self.clientId = clientId
+        self.scheme = scheme
+        self.audience = audience
+        self.oauthTokenGetter = oautTokenGetter
+        self.contextProvider = contextProvider
+    }
+
+    /// Get the user's OAuth token from their Google account. This token can be used to authenticate with the WordPress backend.
+    public func getOAuthToken() async throws -> IDToken {
+        let pkce = ProofKeyForCodeExchange()
+        let url = try await getURL(clientId: clientId, scheme: scheme, pkce: pkce)
+        return try await requestOAuthToken(url: url, clientId: clientId, audience: audience, pkce: pkce)
+    }
+
+    func getURL(clientId: GoogleClientId, scheme: String, pkce: ProofKeyForCodeExchange) async throws -> URL {
+        let url = try URL.googleSignInAuthURL(clientId: clientId, pkce: pkce)
+        return try await withCheckedThrowingContinuation { continuation in
+            let session = ASWebAuthenticationSession(
+                url: url,
+                callbackURLScheme: scheme,
+                completionHandler: { result in
+                    continuation.resume(with: result)
+                }
+            )
+
+            session.presentationContextProvider = contextProvider
+            // FIXME: should this be configurable and/or defaulting to what?
+            session.prefersEphemeralWebBrowserSession = false
```

Comment:
I think for now we can leave it as-is – it wouldn't be too hard to make it configurable, but as a user I can't see a reason I'd want to have to re-login to Google when my Safari is already signed in – that's kind of annoying, especially if I have to 2FA again

Reply:
Fair enough. I might leave a note about it in the code, then.

Reply:
Done via 85a3c271